ENH/TST/DOC: set infer_nrows for read_fwf (GH15138)

rdmontgomery opened the pull request:
- [ ] closes #15138
- [ ] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

`read_fwf` can infer the widths of columns of a dataset from a file or buffer, but if the widths change after the first 100 rows, then the default behavior fails to capture the data correctly. This PR addresses that issue by allowing the user to set a parameter `infer_nrows` that is passed down to the appropriate function.

The associated test ensures that `colspecs` can be determined from the first row, thereby missing data from the subsequent rows and verifying the expected behavior.

Commit:
ENH: set infer_nrows for read_fwf (GH15138)

Previously, read_fwf would infer the width of the columns from the first
100 rows of data. It now accepts a parameter infer_nrows that lets the
user choose how many rows to use in the inference.

Commit:
Add documentation for infer_nrows

Fixed reference in colspecs. Switched half-open interval
representation from non-standard [from, to[ to standard [from, to).

Commit:
Remove documentation for infer_nrows='all'

Commit:
Add test to explicitly set infer_nrows

pep8speaks commented:
Hello @rdmontgomery! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`pandas/io/parsers.py`](https://github.com/rdmontgomery/pandas/blob/da081a868c403ed42ad754adc32333e36bdf06cd/pandas/io/parsers.py) !

 - There are no PEP8 issues in the file [`pandas/tests/io/parser/test_read_fwf.py`](https://github.com/rdmontgomery/pandas/blob/da081a868c403ed42ad754adc32333e36bdf06cd/pandas/tests/io/parser/test_read_fwf.py) !

WillAyd commented on pandas/io/parsers.py:
```diff
@@ -716,7 +720,8 @@ def parser_f(filepath_or_buffer,
 
 @Appender(_read_fwf_doc)
-def read_fwf(filepath_or_buffer, colspecs='infer', widths=None, **kwds):
+def read_fwf(filepath_or_buffer, colspecs='infer', infer_nrows=100,
```
Should add the new keyword to the end of positional arguments so as not to interfere with existing user code

rdmontgomery replied:
Thanks, @WillAyd.

WillAyd commented on pandas/io/parsers.py:
```diff
@@ -3420,11 +3428,11 @@ def get_rows(self, n, skiprows=None):
         return detect_rows
 
-    def detect_colspecs(self, n=100, skiprows=None):
+    def detect_colspecs(self, infer_nrows, skiprows=None):
```
Is there a reason why you changed this from required to optional? Think that might be causing your test failures

rdmontgomery replied:
It was an oversight. Fixed now. Thanks!

Commit:
Move new keyword to end of positional arguments

Commit:
Make infer_nrows optional to retain behavior

Commit:
Make infer_nrows optional when invoking object

rdmontgomery commented:
Hi @WillAyd, thanks for reviewing my code. I made the requested changes and it looks like everything passed now. Is there anything else I can do for this?

WillAyd commented on pandas/io/parsers.py:
```diff
@@ -3361,13 +3367,15 @@ class FixedWidthReader(BaseIterator):
     """
 
-    def __init__(self, f, colspecs, delimiter, comment, skiprows=None):
+    def __init__(self, f, colspecs, delimiter, comment, infer_nrows=100,
```
Still need to change position of `infer_nrows` to be at the end here

WillAyd commented on pandas/io/parsers.py:
```diff
@@ -3420,11 +3428,11 @@ def get_rows(self, n, skiprows=None):
         return detect_rows
 
-    def detect_colspecs(self, n=100, skiprows=None):
+    def detect_colspecs(self, infer_nrows=100, skiprows=None):
```
Same as above

rdmontgomery replied:
Oh, I actually see another problem here. I co-opted `n` because it was performing the intended function of `infer_nrows`, but that would also break previous code that relied on calling `n` explicitly. I'll return the code and tack on `infer_nrows` at the end, like you pointed out.

Commit:
Let infer_nrows default to None to keep behavior

Rework so that if infer_nrows is not set, detect_colspecs still uses
n=100, preserving previous behavior. If infer_nrows comes down the pipe,
the default n is overwritten. Also make sure that the new parameter
infer_nrows is the trailing keyword in all function calls.

Fix the documentation to reflect these changes.

jreback commented on pandas/io/parsers.py:
```diff
@@ -342,14 +342,19 @@
-    A list of field widths which can be used instead of 'colspecs' if
+    the data which are not being skipped via skiprows (default='infer'),
+    or by using the `infer_nrows` parameter.
+infer_nrows : int, default None
```
can you add infer_nrows after widths. pls add a versionadded tag

jreback replied:
default 100

rdmontgomery replied:
I'm switching the places and adding a versionadded tag.

Regarding the default of `None`, I mention my [reasoning in the comment for `read_fwf`](https://github.com/pandas-dev/pandas/pull/23238/files/a35abd8900e9bff3ceb2a81cdd3aecc76375dd35#r228607376). I do state here in the docs
``If not set (or set to `None`), default behavior of 100 rows is used.``

jreback replied:
right but better to simply say its 100

jreback commented on pandas/io/parsers.py:
```diff
@@ -716,7 +722,8 @@ def parser_f(filepath_or_buffer,
 @Appender(_read_fwf_doc)
-def read_fwf(filepath_or_buffer, colspecs='infer', widths=None, **kwds):
+def read_fwf(filepath_or_buffer, colspecs='infer', widths=None,
+             infer_nrows=None, **kwds):
```
shouldn't this be 100?

rdmontgomery replied:
I might be bending over to maintain backwards compatibility here (or I might be wrong), but I decided to let `infer_nrows` default to `None` so that if there is old code that had explicitly called `detect_colspecs(n=1234)`, the function would behave as they expected. If someone sets `infer_nrows` to a number, then `n` is overwritten with `infer_nrows`.

rdmontgomery replied:
To clarify a little further, this was the only way I could think to respect old code that explicitly set `n` in a direct call to `detect_colspecs`. Any new code that chooses to use `infer_nrows` will overwrite `n`, but I did not want to automatically overwrite `n` with a default of 100 if someone had already chosen to do otherwise.

jreback replied:
this is still backward compat. If n is not None, use it, else use infer_nrows. (and let's deprecate setting n that way anyhow).

jreback commented on pandas/tests/io/parser/test_read_fwf.py:
```diff
@@ -141,6 +141,17 @@ def test_fwf_colspecs_None(self):
+  1  2
+123 98
+"""
+        df = read_fwf(StringIO(data), header=None, infer_nrows=1)
```
can you test when infer_nrows > number of rows

rdmontgomery replied:
Sure thing.

jreback reviewed (changes requested):
pls add a whatsnew note as well (other enhancements)

Commit:
Add a whatsnew note

Commit:
Add test for infer_nrows > number of rows

Commit:
Switch places and add versionadded tag

rdmontgomery committed:
Merge branch 'master' into fwf-infer-nrows

rdmontgomery commented on pandas/io/parsers.py:
```diff
@@ -3422,10 +3435,12 @@ def get_rows(self, n, skiprows=None):
         return detect_rows
 
-    def detect_colspecs(self, n=100, skiprows=None):
+    def detect_colspecs(self, n=100, skiprows=None, infer_nrows=None):
```
That's what I had proposed initially and had replaced `n` with `infer_nrows` in place. Your comment about putting the new keyword as a trailing argument is what made me realize that it was possible for someone's code to have instantiated FixedWidthReader and explicitly set `n`. 

Granted, that seems like a rare use case, but I was able to implement a solution that handled both cases. However, this solution seems to have come at the cost of confusing documentation, where `infer_nrows` has a default of ``None`` in order for `read_fwf(colspecs='infer')` to default to inferring from the first 100 rows, as was the case beforehand.

If you and @jreback prefer, I'm happy to use either
`detect_colspecs(self, infer_nrows=100, skiprows=None)` or
`detect_colspecs(self, skiprows=None, infer_nrows=100)`
with the awareness that it might be a breaking change in rare cases.

rdmontgomery commented on pandas/io/parsers.py:
```diff
@@ -343,15 +343,23 @@
     the intervals are contiguous.
+infer_nrows : int, default None
+    The number of rows to consider when letting the parser determine the
+    ``colspecs``. If not set (or set to `None`), default behavior of 100
```
Can you point me to a guide on the use of single and double backticks in pandas or GitHub? I haven't been able to find anything definitive. Or is it just single backticks for code and double backticks for literals? Should colspecs here be in singles?

Commit:
Use infer_nrows unless n is set, default to 100

Preserve backwards compatibility by allowing n to be set, but the new
default will be to use infer_nrows.

Commit:
Merge branch 'fwf-infer-nrows' of https://github.com/rdmontgomery/pandas into fwf-infer-nrows

rdmontgomery commented:
Thanks, guys. I appreciate your help in moving this forward.

I'm not sure what happened with the CircleCI build this time around. Looks like maybe debian.org was having temporary troubles. Any advice here?
> Err:3 http://deb.debian.org/debian stretch Release
>  404  Not Found

rdmontgomery commented:
Hi @jreback and @WillAyd, can you help me understand why the CircleCI test failed? I'd like to re-run the test without pushing any code changes, but I haven't been able to figure out how.

jreback commented:
looks suprious. merge master and push again.

Commit:
Merge branch 'master' into fwf-infer-nrows

WillAyd commented on pandas/io/parsers.py:
```diff
@@ -3421,11 +3433,14 @@ def get_rows(self, n, skiprows=None):
         delimiters = ''.join(r'\%s' % x for x in self.delimiter)
         pattern = re.compile('([^%s]+)' % delimiters)
-        rows = self.get_rows(n, skiprows)
+        if n:
```
Since this is an internal method only and not part of the api I wonder if it wouldn't be better to just replace the `n` keyword with `infer_rows`

rdmontgomery replied:
That sounds reasonable to me, @WillAyd. Would it be better to leave the keyword order as-is with `n` removed or should I reorder like this `def detect_colspecs(self, infer_nrows=100, skiprows=None):`? Or does that really not matter at all?

WillAyd replied:
I'd go with the latter unless @jreback has a differing opinion

jreback replied:
i think ok to reorder that is an internal method

jreback commented:
@rdmontgomery can you merge master. @WillAyd if you have any comments.

WillAyd commented:
@rdmontgomery would be curious if you have thoughts on my previous comment. Always good to simplify / cleanup internal code if possible

jreback commented:
pls merge master as well

jreback committed:
Merge branch 'master' into PR_TOOL_MERGE_PR_23238

jreback commented:
rebased.

rdmontgomery commented:
> 
> 
> @rdmontgomery can you address last comment around the re-ordering of the internal function signature? Once that is done I'm OK to approve and merge

I'll jump on it tomorrow morning! I've been in the last negotiations of a job offer and I just wrapped that up today. I've got some time again!

Commit:
Merge branch 'master' into fwf-infer-nrows

Commit:
Replace internal param with infer_nrows

Let infer_nrows replace the internal parameter n and reorder the
keywords.

Commit:
Add keyword back to function definition

Fix bug introduced when master was merged back in and the function
definition was changed along with the docstring.

WillAyd commented:
Thanks @rdmontgomery !

Good luck with the job
